proof/refine/KHeap_R.thy

```diff
@@ -9,6 +11,59 @@ imports
   ArchMachine_R
 begin
 
+arch_requalify_facts
+  valid_arch_obj'_valid
+
+lemmas [wp] = valid_arch_obj'_valid
+
+(* FIXME: this should go somewhere in spec *)
+translations
+  (type) "'a kernel" <=(type) "kernel_state \<Rightarrow> ('a \<times> kernel_state) set \<times> bool"
+
+declare fromPPtr_def[simp] (* FIXME arch-split: move this somewhere high up *)
+
+(* same derivation on all architectures; some duplication for convenience in simpset *)
+lemma (in Arch) tcb_slots_less_2p_tcbBlockSizeBits:
+  "0 < (2::machine_word) ^ tcbBlockSizeBits"
+  "1 < (2::machine_word) ^ tcbBlockSizeBits"
+  "1 << cteSizeBits < (2::machine_word) ^ tcbBlockSizeBits"
+  "2 << cteSizeBits < (2::machine_word) ^ tcbBlockSizeBits"
+  "3 << cteSizeBits < (2::machine_word) ^ tcbBlockSizeBits"
+  "4 << cteSizeBits < (2::machine_word) ^ tcbBlockSizeBits"
+  "(2::machine_word) ^ cteSizeBits < 2 ^ tcbBlockSizeBits"
+  unfolding tcbBlockSizeBits_def by (auto simp: objBits_simps')
+
+requalify_facts Arch.tcb_slots_less_2p_tcbBlockSizeBits
+declare tcb_slots_less_2p_tcbBlockSizeBits[simp]
+
+(* same derivation on all architectures *)
+lemma (in Arch) zero_one_less_2p_SizeBits[simp]:
+  "(0::machine_word) < 2 ^ cteSizeBits"
+  "(1::machine_word) < 2 ^ cteSizeBits"
+  "(0::machine_word) < 2 ^ epSizeBits"
+  "(1::machine_word) < 2 ^ epSizeBits"
+  "(0::machine_word) < 2 ^ ntfnSizeBits"
+  "(1::machine_word) < 2 ^ ntfnSizeBits"
+  by (auto simp: objBits_simps')
+
+requalify_facts Arch.zero_one_less_2p_SizeBits
+declare zero_one_less_2p_SizeBits[simp]
+
+(* FIXME arch-split: true on all arches, move *)
+lemma zero_less_tcbBlockSizeBits[simp]:
+  "0 < tcbBlockSizeBits"
+  unfolding tcbBlockSizeBits_def by simp
```

Comment:
Is this tagged "move" because it's in between (in Arch) lemmas? if yes, we should probably do the move now, the theory KHeap does otherwise look to be the right place for it (maybe Bits_R, but if it's not needed higher up, KHeap is high enough).

Reply:
I think it was tagged when it was still in the Arch theory, but then after I moved it, it still didn't feel like it was in a good spot. I'll consider moving it to Bits_R. Same for `not_is_other_obj_relation_type` which should go where the other_obj mess lives (and oh man, that name is so bad, I'll take a look too)